devito/arguments.py

```diff
@@ -158,198 +145,398 @@ class PtrArgument(Argument):
+        if self.consumer.name == o.provider.start_name:
+            return 0
 
-    """ Class used to decorate Scalar Function objects with behaviour required for runtime
-        arguments.
+        if o in self.known_values and not isinstance(self.known_values[o],
+                                                     UnevaluatedDependency):
+            provided_values = [self.known_values[o]]
+        else:
+            provided_values = [get_value(o, x, self.known_values)
+                               for x in o.gets_value_from]
+
+        if len(provided_values) > 0:
+            if not all(x is not None for x in provided_values):
+                unknown_args = [x.obj for x in o.gets_value_from
+                                if get_value(o, x, self.known_values) is None]
+
+                def late_evaluate_dim_size(consumer, known_values, partial_values):
+                    known = []
+                    try:
+                        new_values = [known_values[x] for x in unknown_args]
+                        known = [x for x in new_values if x is not None]
+                    except KeyError:
+                        pass
+                    return reduce(max, partial_values + known)
+
+                return UnevaluatedDependency(o, late_evaluate_dim_size,
+                                             [x for x in provided_values
+                                              if x is not None])
+            value = reduce(max, provided_values)
+        else:
+            value = None
+        return value
+
+    def visit_TensorArgument(self, o, param):
+        assert(isinstance(self.consumer, DimensionParameter))
+        return self.known_values[o].shape[param]
+
+
+class Dependency(object):
+    """ Object that represents an edge between two nodes on a dependency graph
+        Dependencies are directional, i.e. A -> B != B -> A
+        :param dependency_type: A dependency can either by of type "gets_value_from"
+               in which case this dependency will be followed for value derivation
+               or it can be of type "verified_by" in which case this dependency will be
+               followed for verification. Both types of dependencies may exist between a
+               pair of nodes. However, only a single dependency of a type may exist
+               between any pair of nodes.
+        :param obj: Dependencies have a source node and a target node. The source node
+                    will store the dependency object on itself. The obj referred to here
+                    is the target node of the dependency.
+        :param param: An optional parameter that might be required to evaluate the
+                      relationship being defined by this Dependency. e.g. when a Dimension
+                      derives its value from a SymbolicFunction's shape, this param
+                      carries the index of a dimension in the SymbolicFunction's shape.
     """
+    GETS_VALUE_FROM = "gets_value_from"
```

Comment:
ooohhh, but there's a hierarchy here! So you probably don't need ``GETS_VALUE_FROM, VERIFIED_BY, _types, ...``, and you can also drop ``self.dependency_type``. This information is already encoded in the type of the subclass. 
The simplest thing to do is to introduce the usual ``is_VerifiedBy/is_GetsValueFrom`` (better name? ``is_GettingValue``?). This way, the subclasses become empty, apart from the presence of ``is_X = True``

Reply:
Done